src/Http/Http.Extensions/gen/StaticRouteHandlerModel/InvocationOperationExtensions.cs

```diff
@@ -11,52 +11,91 @@ namespace Microsoft.AspNetCore.Http.RequestDelegateGenerator.StaticRouteHandlerM
+        "Map",
+        "MapMethods",
+        "MapFallback"
+    };
 
     public static bool TryGetRouteHandlerMethod(this IInvocationOperation invocation, SemanticModel semanticModel, out IMethodSymbol? method)
     {
+        method = null;
+        if (invocation.TryGetRouteHandlerArgument(out var argument))
+        {
+            method = ResolveMethodFromOperation(argument, semanticModel);
+            return true;
+        }
+        return false;
+    }
+
+    public static bool TryGetRouteHandlerArgument(this IInvocationOperation invocation, [NotNullWhen(true)]out IArgumentOperation? argumentOperation)
+    {
+        argumentOperation = null;
+        // Route handler argument is typically the last parameter provided to
+        // the Map methods.
+        var routeHandlerArgumentOrdinal = invocation.Arguments.Length - 1;
         foreach (var argument in invocation.Arguments)
         {
-            if (argument.Parameter?.Ordinal == RouteHandlerArgumentOrdinal)
+            if (argument.Parameter?.Ordinal == routeHandlerArgumentOrdinal)
             {
-                method = ResolveMethodFromOperation(argument, semanticModel);
+                argumentOperation = argument;
                 return true;
             }
         }
-        method = null;
         return false;
     }
 
-    public static IArgumentOperation? GetRouteHandlerArgument(this IInvocationOperation invocation)
+    public static bool TryGetMapMethodName(this SyntaxNode node, out string? methodName)
     {
-        foreach (var argument in invocation.Arguments)
+        methodName = default;
+        if (node is InvocationExpressionSyntax
         {
-            if (argument.Parameter?.Ordinal == RouteHandlerArgumentOrdinal)
+            Expression: MemberAccessExpressionSyntax
             {
-                return argument;
-            }
+                Name: IdentifierNameSyntax
+                {
+                    Identifier: { ValueText: var method }
+                }
+            },
+        })
+        {
+            methodName = method;
+            return true;
         }
-        return null;
+        return false;
     }
 
-    public static bool TryGetRouteHandlerPattern(this IInvocationOperation invocation, out SyntaxToken token)
+    public static bool TryGetRouteHandlerPattern(this IInvocationOperation invocation, [NotNullWhen(true)] out string? token)
     {
-        IArgumentOperation? argumentOperation = null;
-        foreach (var argument in invocation.Arguments)
+        token = default;
+        if (invocation.Syntax.TryGetMapMethodName(out var methodName))
         {
-            if (argument.Parameter?.Ordinal == RoutePatternArgumentOrdinal)
+            if (methodName == "MapFallback" && invocation.Arguments.Length == 2)
             {
-                argumentOperation = argument;
+                token = "{*path:nonfile}";
+                return true;
+            }
+            foreach (var argument in invocation.Arguments)
```

Comment:
Before we were running this code without checking `TryGetMapMethodName`. Is it OK that we are doing this `foreach` inside that check now?

Reply:
Yes, it should be. We shouldn't be entering this codepath if the invocation doesn't have a method name that we support. Since the previous route pattern didn't do the check it ran the risk of producing a value if we called it without verifying it was a known method:

```
app.NotActuallyAMap("a-string-arg-thats-not-a-route-pattern");
```

We do have a top-level check for the known method when we register the generator so this isn't making a functional difference in teh code.